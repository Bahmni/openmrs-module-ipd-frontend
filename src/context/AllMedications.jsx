import React, { useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { ALL_DRUG_ORDERS_URL } from "../constants";

export const AllMedicationsContext = React.createContext({
  getAllDrugOrders: () => {},
});

export const AllMedicationsContextProvider = (props) => {
  const getAllDrugOrders = async (visitUuid) => {
    setMedications((prevState) => ({ ...prevState, isLoading: true }));
    try {
      const response = await axios.get(ALL_DRUG_ORDERS_URL(visitUuid), {
        withCredentials: true,
      });
      if (response.status !== 200) throw new Error(response.statusText);
      setMedications((prevState) => ({
        ...prevState,
        isLoading: false,
        data: response.data,
      }));
    } catch (error) {
      setMedications((prevState) => ({
        ...prevState,
        isLoading: false,
        error,
      }));
    }
  };

  const initialState = {
    isLoading: false,
    data: null,
    error: null,
    getAllDrugOrders,
  };

  const [medications, setMedications] = useState(initialState);

  return (
    <AllMedicationsContext.Provider value={medications}>
      {props.children}
    </AllMedicationsContext.Provider>
  );
};

AllMedicationsContextProvider.propTypes = {
  children: PropTypes.children,
};
