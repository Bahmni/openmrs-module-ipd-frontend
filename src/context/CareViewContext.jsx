import React from "react";

export const CareViewContext = React.createContext();
