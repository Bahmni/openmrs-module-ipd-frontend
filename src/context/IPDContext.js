import React from "react";

export const IPDContext = React.createContext();