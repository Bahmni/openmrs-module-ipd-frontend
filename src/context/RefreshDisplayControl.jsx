import React from "react";

const RefreshDisplayControl = React.createContext({
  refreshDisplayControl: () => {},
});

export default RefreshDisplayControl;
