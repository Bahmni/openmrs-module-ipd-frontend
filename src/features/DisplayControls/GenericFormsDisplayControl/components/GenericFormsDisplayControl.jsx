import React, { useContext, useEffect, useState } from "react";
import PropTypes from "prop-types";
import { IPDContext } from "../../../../context/IPDContext";
import {
  fetchAllConceptsForForm,
  fetchFormTranslations,
} from "../utils/GenericFormsDisplayControlUtils";
import "../styles/GenericFormDisplayControl.scss";
import moment from "moment";
import {
  Link,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from "carbon-components-react";
import { ViewFormModal } from "./ViewFormModal";

const GenericFormsDisplayControl = (props) => {
  const {
    config: { formName: formName },
  } = props;
  const ipdContext = useContext(IPDContext);
  const [formTemplate, setFormTemplate] = useState([]);
  const [formTranslations, setFormTranslations] = useState({});
  const [formEntries, setFormEntries] = useState([]);
  const [openViewFormModal, setOpenViewFormModal] = useState({});
  const headers = [
    {
      key: "Date",
      header: "Date",
    },
    {
      key: "Time",
      header: "Time",
    },
    {
      key: "Provider",
      header: "Provider",
    },
    {
      key: "",
      header: "",
    },
  ];
  const { allFormsSummary = [], allFormsFilledInCurrentVisit = [] } =
    ipdContext;
  const fetchFormConcepts = async (formUuid) => {
    const response = await fetchAllConceptsForForm(formUuid);
    console.log("response 1 ", response);
    if (response?.resources && response?.resources[0]?.value) {
      return JSON.parse(response?.resources[0]?.value);
    }
  };
  const restructureFormConcepts = (formConcepts) => {
    const form = [];
    formConcepts.map((formConcept) => {
      if (formConcept.type === "obsControl") {
        form.push({
          conceptName: formConcept.concept.name,
          translationKey: formConcept.label.translationKey,
          hasGroupMembers: false,
        });
      } else if (formConcept.type === "obsGroupControl") {
        form.push({
          conceptName: formConcept.concept.name,
          translationKey: formConcept.label.translationKey,
          hasGroupMembers: true,
          members: restructureFormConcepts(formConcept.controls),
        });
      }
    });
    return form;
  };

  useEffect(() => {
    const formUuid = allFormsSummary.find(
      (form) => form.name === formName
    )?.uuid;
    fetchFormConcepts(formUuid).then((response) => {
      console.log("response", response);
      const finalResponse = restructureFormConcepts(response.controls);
      setFormTemplate(finalResponse);
    });
    fetchFormTranslations(formName, formUuid).then((response) => {
      setFormTranslations(response.concepts);
    });
  }, []);

  useEffect(() => {
    getEntriesForSelectedForm();
  }, [allFormsFilledInCurrentVisit]);
  const setModalValue = (uuid, value) => {
    setOpenViewFormModal({ ...openViewFormModal, [uuid]: value });
  };

  const getEntriesForSelectedForm = () => {
    const forms = allFormsFilledInCurrentVisit.filter(
      (form) => form.formName === formName
    );
    forms.sort((a, b) => b.encounterDateTime - a.encounterDateTime);
    setFormEntries(forms);
  };
  console.log(
    "form",
    formTemplate,
    formTranslations,
    allFormsFilledInCurrentVisit,
    formEntries
  );
  return (
    <Table useZebraStyles className={"form-table-display-control"}>
      <TableHead>
        <TableRow>
          {headers.map((header, index) => (
            <TableHeader
              key={index + header.key}
              className={`${
                header.key === "Provider" && "provider-column-width"
              }`}
            >
              {header.header}
            </TableHeader>
          ))}
        </TableRow>
      </TableHead>
      <TableBody>
        {formEntries.map((formEntry, index) => {
          const { encounterDateTime, providers } = formEntry;
          const provider = providers[0].providerName;
          return (
            <TableRow key={index}>
              <TableCell>
                {moment(+encounterDateTime).format("DD-MM-YYYY")}
              </TableCell>
              <TableCell>
                {moment(+encounterDateTime).format("HH:mm")}
              </TableCell>
              <TableCell>{provider}</TableCell>
              <TableCell>
                <Link
                  onClick={() => {
                    setModalValue(formEntry.encounterUuid, true);
                  }}
                >
                  View
                </Link>
              </TableCell>
              {openViewFormModal[formEntry.encounterUuid] && (
                <ViewFormModal
                  encounterUuid={formEntry.encounterUuid}
                  callbacks={{ setModalValue }}
                  form={{ formTemplate, formTranslations }}
                  metadata={{
                    formName,
                    encounterDateTime: formEntry.encounterDateTime,
                    provider,
                  }}
                />
              )}
            </TableRow>
          );
        })}
      </TableBody>
    </Table>
  );
};

export default GenericFormsDisplayControl;

GenericFormsDisplayControl.propTypes = {
  patientId: PropTypes.string,
  config: PropTypes.object,
};
