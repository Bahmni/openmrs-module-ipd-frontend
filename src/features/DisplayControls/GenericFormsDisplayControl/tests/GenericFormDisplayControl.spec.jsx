import React from "react";
import { render, waitFor } from "@testing-library/react";
import GenericFormsDisplayControl from "../components/GenericFormsDisplayControl";
import { IntlProvider } from "react-intl";
import { IPDContext } from "../../../../context/IPDContext";
import {
  allFormsFilledInVisitMock,
  allFormsSummaryMock,
  mockFormConcepts,
  mockFormTranslations,
} from "./GenericFormDisplayControlMock";

jest.mock("../utils/GenericFormsDisplayControlUtils", () => {
  const originalModule = jest.requireActual(
    "../utils/GenericFormsDisplayControlUtils"
  );
  return {
    ...originalModule,
    fetchAllConceptsForForm: jest.fn(() => {
      return Promise.resolve(mockFormConcepts);
    }),
    fetchFormTranslations: jest.fn(() => {
      return Promise.resolve(mockFormTranslations);
    }),
  };
});
describe("Generic Form Display Control", () => {
  it("should match snapshot", async () => {
    const config = {
      formName: "Test Form",
    };
    const { container, getAllByText } = render(
      <IPDContext.Provider
        value={{
          allFormsSummary: allFormsSummaryMock,
          allFormsFilledInCurrentVisit: allFormsFilledInVisitMock,
          config: { enable24HourTime: true },
        }}
      >
        <IntlProvider locale={"en"}>
          <GenericFormsDisplayControl config={config} />
        </IntlProvider>
      </IPDContext.Provider>
    );
    await waitFor(() => {
      expect(getAllByText("Mock Provider")).toBeTruthy();
    });
    expect(container).toMatchSnapshot();
  });
});
