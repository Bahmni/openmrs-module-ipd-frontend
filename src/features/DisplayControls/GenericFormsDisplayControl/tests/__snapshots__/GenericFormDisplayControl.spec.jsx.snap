// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic Form Display Control should match snapshot 1`] = `
<div>
  <div
    class="bx--data-table-content"
  >
    <table
      class="bx--data-table form-table-display-control bx--data-table--zebra bx--data-table--no-border"
    >
      <thead>
        <tr>
          <th
            class="false"
            scope="col"
          >
            <div
              class="bx--table-header-label"
            >
              Date
            </div>
          </th>
          <th
            class="false"
            scope="col"
          >
            <div
              class="bx--table-header-label"
            >
              Time
            </div>
          </th>
          <th
            class="provider-column-width"
            scope="col"
          >
            <div
              class="bx--table-header-label"
            >
              Provider
            </div>
          </th>
          <th
            class="false"
            scope="col"
          />
        </tr>
      </thead>
      <tbody
        aria-live="polite"
      >
        <tr>
          <td>
            24-04-2024
          </td>
          <td>
            10:40
          </td>
          <td>
            Mock Provider
          </td>
          <td>
            <a
              class="bx--link"
            >
              View
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
